src/SFA.DAS.AssessorService.Application/Handlers/Approvals/ImportApprovalsHandler.cs

```diff
@@ -0,0 +1,120 @@
+            , IApprovalsExtractRepository approvalsExtractRepository
+            , ISettingRepository settingRepository
+            , IOuterApiService outerApiService
+            )
+        {
+            _logger = logger;
+            _approvalsExtractRepository = approvalsExtractRepository;
+            _settingRepository = settingRepository;
+            _outerApiService = outerApiService;
+        }
+
+        public async Task<Unit> Handle(ImportApprovalsRequest request, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _logger.LogInformation("Commencing approvals import.");
+
+                // 1. Figure out extract start time.
+
+                DateTime? extractStartTime = null;
+                DateTime? latestApprovalsExtractTimestamp = await _approvalsExtractRepository.GetLatestExtractTimestamp();
+                if(null != latestApprovalsExtractTimestamp && latestApprovalsExtractTimestamp.HasValue)
+                {
+                    extractStartTime = latestApprovalsExtractTimestamp.Value.AddSeconds(- await GetSettingAsInt(TOLERANCE_SETTING_NAME));
+                }
+
+                // 2. Request the extract in batches.
+
+                int batchSize = await GetSettingAsInt(BATCHSIZE_SETTING_NAME);
+                int batchNumber = 0;
+                int count = 0;
+                GetAllLearnersResponse learnersBatch = null;
+
+                do
+                {
+                    batchNumber++;
+                    learnersBatch = await _outerApiService.GetAllLearners(extractStartTime, batchNumber, batchSize);
+                    if (null == learnersBatch || null == learnersBatch.Learners)
+                    {
+                        throw new Exception($"Failed to get learners batch: sinceTime={extractStartTime?.ToString("o", System.Globalization.CultureInfo.InvariantCulture)} batchNumber={batchNumber} batchSize={batchSize}");
+                    }
+
+                    // 3. Upsert ApprovalsExtract batch.
+
+                    UpsertApprovalsExtract(learnersBatch.Learners);
+                    count += learnersBatch.Learners.Count;
+
+                } while (batchNumber < learnersBatch.TotalNumberOfBatches);
+
+                // 4. Run Populate Learner
+
+                var learnerCount = await _approvalsExtractRepository.PopulateLearner();
+
+                _logger.LogInformation($"Approvals import completed successfully. {count} record(s) read from outer api, {learnerCount} records inserted to Learner table.");
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Approvals import failed to complete successfully.");
+                throw;
+            }
+
+            return Unit.Value;
+        }
+
+        private async Task<int> GetSettingAsInt(string name)
+        {
+            int value = 0;
+
+            var settingValue = await _settingRepository.GetSetting(name);
```

Comment:
What was the thinking behind the _settingRepository and the database over the EmployerConfig that gets injected and adding a setting?

Reply:
It was a requirement in the JIRA / Wiki page - "Note, the Settings table, to hold name value pairs, in Assessor Db das-prd-ass-db is currently used by the ILR update processing. This will enable soft configuration of the batch size and time overlap for the /api/learners call - and will require a post-deployment script to configure at deployment time."   [](https://skillsfundingagency.atlassian.net/wiki/spaces/NDL/pages/3133210746/Approvals+database+-+SV#Assessor-sevice-API-to-Update-Approvals-Extract)